Audit and merge back `feature/amk_missions` into `main`

zach2good opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

I was reminded that this was languishing to the side in a feature branch. There have been a lot of architectural changes while this has been alive, so it'll take some massaging to get into a state that can merge.

Each PR that has gone into here has been reviewed by me, so there shouldn't really be any surprises.

zach2good committed:
Start latter AMK missions

zach2good committed:
Progress on AMK Cardian Fight

Flibe-XI committed:
[AMK] Update helm logic, added chocobo digging sites (#4579)

Flibe-XI committed:
added reminder text and repeatability to amk6

Flibe-XI committed:
technically correct cs mission order

Flibe-XI committed:
mission 9 complete

Flibe-XI committed:
[AMK, LUA] mission 12 - Castle Zvahl cutscene (#4935)

Flibe-XI committed:
[AMK][Lua] Cardian orb drop mechanic (#4800)

* AMK6: Added mechanic to drop KI orbs for cardians

* Fixed null player check

Flibe-XI committed:
[AMK, Lua] Mission 13 puzzle 1 - Beaucedine Glacier runaround (#5045)

* Add beauc glac puzzle npc logic

Structure for cutscenes handled by lonely evergreen and goblin grenadier, as well as the
structure of the pips.

* puzzle pips complete, map bug fixed

The cutscene event for the goblin grenadier requires two sets of args, first is for the hint and win/lose mechanic, the second
from the eventUpdate section is to relabel an pips on the map if you wanted to.  Setting the pip args to 0 makes it so the
map isn't changed accidentally

Flibe-XI committed:
puzzle 3 triva challenge implementation

Almost all logic for determining cutscenes is handled by event arguments
I've done my best to make all the arguments self-explanatory via their names

Flibe-XI committed:
Finished out questions, fixed some logic errors

Also condensed functions for beauc puzzle into helpers.lua

Flibe-XI committed:
Implement Castle Zvahl run

Flibe-XI committed:
puzzle 4 implemented

Various bug fixes, and repeatability added to lonely evergreen, whereas
shadowy pillar is not needed to be talked to again after losing battle

MowFord commented:
If you would like me or @Flibe-XI to help squash conflicts I don't mind (can't speak for him)

zach2good committed:
AMK: Restore missing function after rebase

zach2good commented:
> If you would like me or `Flibe-XI` to help squash conflicts I don't mind (can't speak for him)

No need, rebase only took like 5 mins 👍

zach2good commented:
TODO: Just need to run UpdateExtractor to make sure the text IDs aren't shifted (some of them are certainly busted), then this is good for a test runthrough

zach2good commented:
TODO: Reconcile differences between AMK and base helm: https://github.com/LandSandBoat/server/pull/4579

zach2good commented:
IDs look good vs `main`, `Sanity_Checks` passing, etc.

claywar commented on scripts/missions/amk/01_A_Moogle_Kupo_dEtat.lua:
```diff
@@ -3,6 +3,10 @@
 -- A Moogle Kupo d'Etat M1
 -- !addmission 10 0
 -----------------------------------
+require('scripts/globals/missions')
```
Global requires are no longer necessary outside of dependencies within globals alone (not going to tag every instance)

claywar commented on scripts/missions/amk/06_An_Errand_The_Professors_Price.lua:
```diff
@@ -20,28 +25,36 @@ mission.reward =
     local removedKIs = 0
-    for _, keyItemId in ipairs(orbKeyItems) do
+    for _, entry in ipairs(orbKeyItems) do
+        local keyItemId   = entry[1]
```
I wonder, could this be simplified to `local keyItemId, immunity, immuneValue, vulnValue = unpack(entry)`?

Flibe-XI replied:
Yes, however I believe xi.mod.BLUNT_SDT needs to be replaced with IMPACT and HTH, so the table itself needs to be reworked.  I think @Xaver-DaRed is working on this so I'll leave it alone.

claywar commented on scripts/missions/amk/06_An_Errand_The_Professors_Price.lua:
```diff
@@ -56,18 +69,33 @@ local beginCardianFight = function(player, npc)
         check = function(player, currentMission, missionStatus, vars)
-            return currentMission == mission.missionId and not player:hasKeyItem(xi.ki.RIPE_STARFRUIT)
+            return currentMission >= mission.missionId and
+            missionStatus == 0 and
```
need indentation for lines after the return

claywar commented on scripts/missions/amk/06_An_Errand_The_Professors_Price.lua:
```diff
@@ -84,14 +112,27 @@ mission.sections =
                 onTrigger = function(player, npc)
-                    return mission:progressEvent(100)
+                    if
+                        player:hasKeyItem(xi.ki.ORB_OF_SWORDS) or
```
Would almost prefer we iterate over orbKeyItems ([1]), but not overly large

claywar commented on scripts/missions/amk/06_An_Errand_The_Professors_Price.lua:
```diff
@@ -103,7 +144,8 @@ mission.sections =
     {
         check = function(player, currentMission, missionStatus, vars)
-            return currentMission == mission.missionId and
+            return currentMission >= mission.missionId and
```
This seems dangerous since we're allowing status of 0 for any and all missions including or greater than 6

claywar commented on scripts/missions/amk/09_Rescue_A_Moogles_Labor_of_Love.lua:
```diff
@@ -0,0 +1,208 @@
+    nextMission = { xi.mission.log_id.AMK, xi.mission.id.amk.ROAR_A_CAT_BURGLAR_BARES_HER_FANGS },
+}
+
+local markerSets =
```
Perhaps a comment for the intent of this matrix?

claywar commented on scripts/missions/amk/09_Rescue_A_Moogles_Labor_of_Love.lua:
```diff
@@ -0,0 +1,208 @@
+end
+
+local hasAllStones = function(player)
+    for offset = 0, 8 do
```
The loop would probably be more clear instead of referring to offset, but iterating `for keyItemId = xi.ki.STONE_OF_SURYA, xi.ki.<whatever_is_end>`

claywar commented on scripts/missions/amk/09_Rescue_A_Moogles_Labor_of_Love.lua:
```diff
@@ -0,0 +1,208 @@
+    {
+        check = function(player, currentMission, missionStatus, vars)
+            return currentMission == mission.missionId and
+            player:getCharVar('Mission[10][8]progress') == 0
```
Not sure why a charvar is being used here, or why we could be on mission 9 when mission 8 is still has status set

Flibe-XI replied:
Because the talisman is lost upon entry to fight, so if player loses, they have to regain the talisman.  Is there another way I'm not aware of for tracking status of a mission the player isn't currently on? Or should the code for getting the talisman be replicated in the mission 10 file?

claywar commented on scripts/missions/amk/09_Rescue_A_Moogles_Labor_of_Love.lua:
```diff
@@ -0,0 +1,208 @@
+
+                    -- Determine if QM triggered is in markerset
+                    local keyItem = 0
+                    for setIndex = 1, 9 do
```
If this is the entire size of the matrix, prefer ipairs

claywar commented on scripts/missions/amk/09_Rescue_A_Moogles_Labor_of_Love.lua:
```diff
@@ -0,0 +1,208 @@
+            onEventFinish =
+            {
+                [100] = function(player, csid, option, npc)
+                    for i = 0, 8 do
```
Similar to above, no reason for an offset, just iterate through the keyitem IDs

Flibe-XI commented:
> ID updates run, amk_helpers & helm changes reconciled - ready for testing if anyone has time (I'm still slammed at work)

I will give it a go in the next few days.

Flibe-XI committed:
Removed requires from header

Flibe-XI committed:
Added comment describing table in mission 9

Changed loop from range to enums

Flibe-XI committed:
Fix: corrected helmtype in amk helpers.lua

claywar commented on scripts/missions/amk/07_Shock_Arrant_Abuse_of_Authority.lua:
```diff
@@ -6,8 +6,6 @@
 -- Note: KI aquisition is handled in chocobo_digging.lua
```
Disregard previous comment, wrong mission lines during a late night

zach2good commented:
I've rebased and cleaned up any ID collisions etc. This all looks OK and passes sanity checks, once the other CI groups pass I'll merge this, then it's only the BCNMs to do? Which would want to be in the new BCNM file style.

There may or may not be useful bits to take from my old DSP PR: https://github.com/DarkstarProject/darkstar/pull/6053
